### PR TITLE
Add codebase overview reference

### PR DESCRIPTION
**Pull request by mohavro**
## Summary
- add a CODEBASE_OVERVIEW.md to consolidate architecture, conventions, and workflows for new contributors
- capture backend, frontend, data flow, testing, and extension guidance in a single reference document

## Testing
- Not run (not requested)

[Codex Task](https://chatgpt.com/codex/tasks/task_e_693dae6b2a088324bbd42cf6982f7ead)

### **Auto-created Ticket**

[#462](https://github.com/DashFin/financial-asset-relationship-db/issues/462)

### **PR Type**
Documentation

### **Description**
- Add comprehensive CODEBASE_OVERVIEW.md reference document

- Consolidate architecture, conventions, and workflows for contributors

- Document system purpose, entry points, and high-level architecture

- Provide guidance on domain models, backend conventions, and extension patterns

### Diagram Walkthrough

```mermaid
flowchart LR
  A["New Contributors"] -->|"reads"| B["CODEBASE_OVERVIEW.md"]
  B -->|"explains"| C["Architecture & Design"]
  B -->|"explains"| D["Coding Standards"]
  B -->|"explains"| E["Testing & Deployment"]
  B -->|"explains"| F["Extension Patterns"]
  C -->|"covers"| G["Frontends, Backend, Core Logic"]
  D -->|"covers"| H["Python, Libraries, File Conventions"]
  E -->|"covers"| I["Pytest, Jest, Docker, Vercel"]
  F -->|"covers"| J["New Assets, Relationships, Visualizations"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CODEBASE_OVERVIEW.md</strong><dd><code>Comprehensive codebase overview and contributor reference</code></dd></summary>
<hr>

CODEBASE_OVERVIEW.md

<ul><li>New comprehensive reference document for repository architecture and <br>conventions<br> <li> Documents system purpose, entry points, and high-level architecture <br>overview<br> <li> Covers domain models, relationship patterns, and backend service <br>conventions<br> <li> Provides data flow, coding standards, testing procedures, and <br>extension guidance<br> <li> Includes deployment notes and quick-start development instructions</ul>

</details>

  </td>
  <td><a href="https://github.com/DashFin/financial-asset-relationship-db/pull/461/files#diff-f4893b63d6a298a69f1deb8733a1c74c70f5ea1be04fec45dfef6eb4ea470d43">+62/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

## **CodeAnt-AI Description**
Add comprehensive tests for auth and PR context processing; add codebase overview and test reports

### What Changed
- Added extensive unit tests covering authentication (JWT, password hashing, user repo), PR context processing/token counting, and many SQLite/database edge cases; these tests exercise happy paths, edge cases, and error handling.
- Added CODEBASE_OVERVIEW.md summarizing repository architecture, entry points, conventions, testing and deployment guidance for new contributors.
- Added companion test-generation documentation and quick-start guides that list new test files, how to run them, and coverage expectations.

### Impact
`✅ Fewer regressions in authentication and database behavior`
`✅ Shorter onboarding for new contributors`
`✅ Faster PR review with runnable, targeted tests`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: CODEBASE_OVERVIEW.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+- Relationship discovery emphasizes both **bidirectional** (e.g., same sector) and **directional** (e.g., corporate bond → equity) links with deterministic 3D layouts using fixed seeds so visual output remains stable.
+- Visualization inputs are normalized: graph nodes/edges are colored by asset class, node sizes scale by importance, and metrics include relationship density and top relationships.
+## Backend Service Conventions
+- Graph lifecycle: global graph instance is initialized lazily and guarded by a threading lock; `set_graph_factory` enables test overrides. Real-data mode is toggled via environment flags (`USE_REAL_DATA_FETCHER`, cache path variables), falling back to sample data when disabled.
+- Authentication: `/token` issues JWTs using username/password credentials; all protected routes depend on the authenticated user model. Access tokens expire after 30 minutes by default. JWT signing configuration (e.g., secret key and algorithm) is provided via environment variables/secret management and must not be committed to the repository.
```
Comment (mohavro) [inline CODEBASE_OVERVIEW.md]: @qodo-merge-pro replace line 35 with:

- Authentication: `/token` issues JWTs using username/password credentials. Access tokens expire after 30 minutes and can be renewed using refresh tokens to maintain user sessions. All protected routes depend on a valid access token.

File: tests/unit/test_database.py [hunk 2]
```diff
@@ -356,3 +367,312 @@ def _cause_integrity_error() -> None:
...
+        """Test fetch_one returns None when no results."""
+        mock_conn = Mock()
+        mock_cursor = Mock()
+        mock_cursor.fetchone.return_value = None
+        mock_conn.execute.return_value = mock_cursor
+        mock_get_conn.return_value.__enter__.return_value = mock_conn
+
+        result = fetch_one("SELECT * FROM nonexistent")
+
+        assert result is None
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_database.py]: @qodo-merge-pro fix this

File: tests/unit/test_auth.py [hunk 3]
```diff
@@ -0,0 +1,764 @@
...
+        assert user.disabled is False
+
+    def test_user_model_optional_fields(self):
+        """Test User model with optional fields as None."""
+        user = User(username="testuser")
+
+        assert user.username == "testuser"
+        assert user.email is None
+        assert user.full_name is None
+        assert user.disabled is False
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]: @qodo-merge-pro fix this

File: tests/unit/test_context_chunker.py [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+limits:
+  fallback:
+    priority_order:
+      - review_comments
+      - test_failures
+      - changed_files
+""",
+    )
+    @patch("pathlib.Path.exists")
+    def test_initialization_with_valid_config(self, mock_exists, mock_file):
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_context_chunker.py]: @qodo-merge-pro fix this

File: tests/unit/test_database.py [hunk 5]
```diff
@@ -29,6 +38,8 @@
     session_scope,
 )
 
+pytest.importorskip("sqlalchemy")
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_database.py]: @qodo-merge-pro fix this

File: tests/unit/test_auth.py [hunk 6]
```diff
@@ -0,0 +1,764 @@
...
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]: @qodo-merge-pro fix this

File: tests/unit/test_auth.py [hunk 7]
```diff
@@ -0,0 +1,764 @@
...
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]: @qodo-merge-pro fix this

File: tests/unit/test_auth.py [hunk 8]
```diff
@@ -0,0 +1,764 @@
...
+class TestGetCurrentActiveUser:
+    """Test suite for get_current_active_user dependency."""
+
+    def test_get_current_active_user_active_user(self):
+        """Test with active (non-disabled) user."""
+        active_user = User(username="activeuser", email="active@example.com", disabled=False)
+
+        result = get_current_active_user(active_user)
+
+        assert result == active_user
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]: @qodo-merge-pro fix this

File: tests/unit/test_auth.py [hunk 9]
```diff
@@ -0,0 +1,764 @@
...
+This module provides extensive test coverage for:
+- UserRepository operations (get_user, has_users, create_or_update_user)
+- Password hashing and verification
+- JWT token creation and validation
+- User authentication flows
+- Environment-based credential seeding
+- Edge cases and error conditions
+"""
+
+import os
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]: @qodo-merge-pro fix this

File: tests/unit/test_auth.py [hunk 10]
```diff
@@ -0,0 +1,764 @@
...
+
+        # But cannot access protected resources
+        with pytest.raises(HTTPException) as exc_info:
+            get_current_active_user(retrieved_user)
+        assert exc_info.value.status_code == 400
+
+
+@pytest.mark.unit
+class TestAuthenticationEdgeCases:
+    """Test edge cases and error conditions."""
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]: @qodo-merge-pro fix this

**Comment by mohavro** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/dashfin.fardb/pr-checks/936b4041-2dd2-453a-b6fc-54c3814b0ee5?refs=a35a9102b85f686439c9e18ea1dae0cd2f6970f7&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: tests/unit/test_auth.py [hunk 11]
```diff
@@ -0,0 +1,765 @@
...
+- Environment-based credential seeding
+- Edge cases and error conditions
+"""
+
+import os
+from datetime import timedelta
+from unittest.mock import Mock, patch
+
+import pytest
+from jose import jwt
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_auth.py]:
```suggestion
jose_module = pytest.importorskip("jose", reason="python-jose is required for auth JWT unit tests")
jwt = jose_module.jwt

from api.auth.import (
    ALGORITHM,
    SECRET_KEY,

)       
```

**Comment by mohavro** [comment]
Closing — cherry-picked the valuable additions manually:

**test_database.py**: Added 7 new test classes for `api/database.py` (`TestResolveSqlitePathEnhancements`, `TestIsMemoryDbEnhancements`, `TestExecuteFunctionEnhancements`, `TestFetchOperationsEnhancements`, `TestDatabaseErrorHandling`, `TestDatabaseUrlConfiguration`) plus required imports (`sqlite3`, `Mock`, `MagicMock`, api.database functions). Skipped `TestConnectionManagementEnhancements` (design flaw: patches `DATABASE_PATH` but `_db_manager` uses its own internal copy). Fixed `test_get_database_url_missing_env_var` — removed invalid `@patch("api.database.DEFAULT_DATABASE_URL")` which doesn't exist in that module.

**test_context_chunker.py**: Created as a new file. Fixed a critical `NameError` bug from the PR — the original used `ContextChunker` in test bodies without it being in scope (the fixture was never declared as a parameter). Fixed by doing a module-level `sys.path` insert + direct import instead.

Skipped: `test_auth.py` (duplicate class, typo in class name, double-dot patch path). Skipped all 4 markdown documentation files.
